### ISSUE TITLE
Request: Remove 'Unload or Reload' Warning/Question (and always Reload)

### ISSUE THREAD

**Issue by gaiking-uk** [AutoHotkey/Ahk2Exe#75]
### Request ###
Hi guys, I would like to suggest removing the warning message that appears when trying to (re)-compile a script that is currently running, and the related question whether the file should be Unloaded of Reloaded.

### Change ###
In this situation, ahk2exe could instead simply **always** (Unload, compile, and then) **Reload** the script.

### Rationale / Context ###
I have a main .ahk script and several supporting .ahk scripts that are #include(d) -- these are compiled into my ahk.exe program which autoruns when I login (and so is always running).

I often make changes to the supporting .ahk scripts (e.g. one of which is the AHK-provided Hotstring Manager / self-appending script) and (unless I missed something) need to re-compile my main .ahk script in order for changes to be pulled through / tested (closing/re-opening the .exe is not sufficient). Sometimes (in frequent change periods) I could get the Unload/Reload prompt 20+ times, which became quite distracting/laborious.

In the end, I removed the in-use check from ahk2exe.ahk (and instead always ran the code to Reloaded the file after it had re-compiled) and re-compiled ahk2exe... Personally, I much prefer the silent-Reload behaviour and it makes updating (and re-compiling my .ahk scripts) much easier now - so I thought I would at least propose the change to you guys (to make globally) if you wanted to do this.

Hopefully the above is clear (and you agree it would be a good change), please let me know if you have any questions/comments.
___
PS1 -- Hopefully this request is taken in the right way, but just in case it seems I am being negative of ahk2exe, please note that is not my intention at all!.. The ahk2exe tool is great, I simply wanted to share a change that has helped me quite a lot personally, so this could be considered for wider roll-out to wider users.

PS2 -- I realise I may not need to re-compile the script after every change and I could just run the un-compiled .ahk script instead, however I sometimes found differences between how the scripts sometimes ran as .ahk files vs. .exe files (e.g. with Hotstrings.ahk, trying to append to the script if it's an .exe will fail, etc).. Also, as the compiled .exe autoruns on load, this meant I was running v1.0 of the .exe and v1.0.1 (i.e. different code) in the .ahk simultaneously, which had unpredictable results... So opting to always re-compile the script, seemed like a reasonable way to overcome these nuances.

**Comment by TAC109** [comment]
I appreciate that the 'reload' prompt might be annoying in your situation, but for other users, to simply unload and reload the compiled .exe without a prompt could lead to very undesirable consequences. 

Did you consider creating a small script that detects the 'reload' prompt from your compile and pushes the appropriate button? (This is AutoHotkey after all!)

**Comment by gaiking-uk** [comment]
Hi, sorry for taking a few days to get back to you - my bad!

**RE: Prompt useful for others** -- Fair enough, I can certainly appreciate that. (FYI) when writing my original request, I did try to validate it was reasonable (i.e. think of any situations where I might not want it to reload automatically) and posted the request as I couldn't think of any situations where I wouldn't want it to reload the script, however I accept that other users may be using AHK in different ways / have scripts where Reloading might be a problem 👍🏼

**RE: Using AHK to detect Ahk2Exe and auto-press Reload button** -- I like the idea (like you say, there is a certain charm in using AHK itself for this), unfortunately thought, interacting with another GUI program is one part of AHK that I have never been able to crack (have tried in the past, as wanted to write a script that auto-pressed menu options in VLC.exe but found the examples online a bit hard to follow/adapt and couldn't get it working). Also though, while I like the idea - it might be a bit hard for general/other users to replicate (i.e. the reason I posted in the first place was to try and contribute something / make the app better for users overall).

In hindsight, I would like to amend my request...
___
### REVISED REQUEST ###
Is it possible to add an option to Ahk2Exe to control the "Existing Script" behaviour (pre-set the behaviour to one of the following options)?
1. `Ask` (default if not specified / as is now)
2. `Reload` (don't prompt, always Reload the script)
3. `Unload` (don't prompt, always Unload the script)

#### Possible option 1: Command line switch ####
`Ahk2Exe /ScriptExists=Reload`

#### Possible option 2: Ahk.Exe ini file ####
```
AHK2EXE.ini
-----------

ScriptExists=Reload
```

Would either of these options be workable?

**Comment by joedf** [comment]
Possibly covered by https://github.com/AutoHotkey/Ahk2Exe/pull/77
if we do decide to merge it.

**Comment by gaiking-uk** [comment]
> Possibly covered by #77
> if we do decide to merge it.

@joedf - Yes, that change would cover this, thanks! 👍🏼... and is a change coming soon I believe(?) (from what I understood on that page, albeit based on limited knowledge of github processes/flows admittedly)

> That's one example where you'd be better off not compiling at all. But what benefit is there to compiling?
@Lexikos - Yes, it was due to portability as was using that AHK script on multiple machines (some of which didn't have AHK installed)

> Were they meant to run simultaneously?
No, my own process/rule was that I should only run the compiled .exe (that's what ran on start-up), but occasionally I would double click the .ahk version instead

> If you're recompiling often, you should be making it easy for yourself by automating the process.
Totally agree (am all about automation) and that's what I did really (in effect I hard-coded the `/forceReload` behaviour into my local branch of `Ahk2Exe`)

> If you use a script to invoke `Ahk2Exe`, you can just have it terminate your compiled script first, and restart it afterward.
Thanks for the tip. (FYI) I made the change to `Ahk2Exe.ahk` a while ago (~1 year), and my knowledge of scripting (generally) was a lot lower back then (didn't know AHK as well, or know much about shell commands like `taskkill`, etc)... Agree that if I was making the change today (with my current level of knowledge), I would probably do it the way you suggested instead 👍🏼

**Comment by Lexikos** [comment]
> Yes, it was due to portability as was using that AHK script on multiple machines (some of which didn't have AHK installed)

It seems to me that by running the compiled version on the same system you use to write and edit it, you are just creating extra unnecessary work for yourself.

It also seems to me that compiling isn't necessarily simpler or easier for portability, when the author is the user.

> [Portability of AutoHotkey.exe](https://www.autohotkey.com/docs/Program.htm#portability)
> 
> The file AutoHotkey.exe is all that is needed to launch any .ahk script.
> 
> [v1.0.90+]: Renaming AutoHotkey.exe also changes which script it runs by default, which can be an alternative to compiling a script for use on a computer without AutoHotkey installed. For instance, MyScript.exe automatically runs MyScript.ahk if a filename is not supplied, but is also capable of running other scripts.

But perhaps that's all beside the point. It doesn't matter why the script is being compiled, or why it would already be running at the time.

**Comment by gaiking-uk** [comment]
> But perhaps that's all beside the point. It doesn't matter why the script is being compiled...

No worries, is interesting that you mentioned that though, as admittedly while I was writing my reply to your points I did think to myself "I'm not sure I'm adding anything here", but is just part of my nature that where possible I like to try and make the effort to respond to people's advice, comments, etc (rather than just leave them un-responded/acknowledged).

> The file AutoHotkey.exe is all that is needed to launch any .ahk script.

OK, so have to admit I hadn't clocked this at all, so thanks for sharing as is really useful to know!

(FYI #1) - As AHK comes with an _installer*_ (and hooks into Windows, the keyboard/mouse etc), I assumed to do all this cool stuff, it must be a non-trivial install (i.e. deploying dll's, services, reg entries, and what-not)... I must admit, I never even thought to try and "pull" `AutoHotkey.exe` out of `C:\Program Files`
(FYI #2) - I didn't know either about the using AutoHotKey.exe as a host/to run an .ahk script. I think the first time I ran AutoHotKey.exe (admittedly years ago now), I saw the homepage, my brain immediately went "oh, it's a just shell app for the website / .chm file", closed it and never thought to use it again)... Checking again today, I can see you guys even added a "Why am I seeing this?" section at the bottom of the page, so sorry I missed this!

With the above additional facts in hand, I can see a lot more clearly where you comments come from / why your suggestions make sense and why my approach might have seemed strange... It's purely just that _at the time_ I assumed (a) AHK had to be installed on the machine in order to run an .AHK script (and if this couldn't be done, the workaround was to compile the .ahk to an .exe)... which led me to down the route of always using/testing the .exe (compiled version) as like I say, things like the WIN+H [Hotstrings Helper](https://www.autohotkey.com/docs/Hotstrings.htm#Helper) didn't work (without adjustment) when compiled, so in essence to make sure any scripts that I wrote would work on both main/dev PC + synced/non-AHK machines, it was best to do everything against the compiled script.

Anyway, I realise I've moved the conversation rather off topic, but you have given me a number of useful points / things to think about, thanks 👍🏼

**Comment by joedf** [comment]
> 
> 
> > Possibly covered by #77
> > if we do decide to merge it.
> 
> @joedf - Yes, that change would cover this, thanks! 👍🏼... and is a change coming soon I believe(?) (from what I understood on that page, albeit based on limited knowledge of github processes/flows admittedly)

In that case, I'll link to the PR to automatically close this if/when merged.
